File: erlang/src/lexer.erl
-module(tokenizer).
-export([create_token/2, is_letter/1, is_number/1, tokenizer/1]).

-record(token, {type, literal}).

create_token(Type, Literal) ->
    #token{type = Type, literal = Literal}.

is_letter(Character) ->
    case Character of
        $a..$z; $A..$Z; $_ -> true;
        _ -> false
    end.

is_number(Character) ->
    case Character of
        $0..$9 -> true;
        _ -> false
    end.

tokenizer(Input) ->
    tokenizer(1, 1, Input, []).

tokenizer(Position, ReadPosition, Input, Tokens) ->
    case read_char(ReadPosition, Input) of
        {Char, _} when Char == $=; Char == $+; Char == $,; Char == $; Char == $($; Char == $); Char == ${; Char == $} ->
            Token = create_token(atom_to_list(type_from_char(Char)), [Char]),
            tokenizer(ReadPosition + 1, ReadPosition + 1, Input, [Token | Tokens]);

        {Char, _} when is_letter(Char) ->
            Ident = read_ident(Position, ReadPosition, Input),
            Token = case keyword_from_ident(Ident) of
                         undefined -> create_token(ident, Ident);
                         Keyword -> Keyword
                     end,
            tokenizer(Position, ReadPosition, Input, [Token | Tokens]);

        {Char, _} when is_number(Char) ->
            Int = read_int(Position, ReadPosition, Input),
            Token = create_token(int, Int),
            tokenizer(Position, ReadPosition, Input, [Token | Tokens]);

        {_, _} -> % Illegal character
            Token = create_token(illegal, [Char]),
            tokenizer(ReadPosition, ReadPosition, Input, [Token | Tokens]);

        eof ->
            lists:reverse([create_token(eof, "eof") | Tokens])
    end.

read_char(Position, Input) ->
    case Position =< length(Input) of
        true -> {element(Position, Input), Position + 1};
        false -> eof
    end.

read_ident(Position, ReadPosition, Input) ->
    case read_char(ReadPosition, Input) of
        {Char, _} when is_letter(Char) ->
            read_ident(Position, ReadPosition + 1, Input);
        _ ->
            string:substr(Input, Position, ReadPosition - Position)
    end.

read_int(Position, ReadPosition, Input) ->
    case read_char(ReadPosition, Input) of
        {Char, _} when is_number(Char) ->
            read_int(Position, ReadPosition + 1, Input);
        _ ->
            string:substr(Input, Position, ReadPosition - Position)
    end.

type_from_char($=) -> equal;
type_from_char($) -> plus;
type_from_char($,) -> comma;
type_from_char($;) -> semicolon;
type_from_char($() -> lparen;
type_from_char($)) -> rparen;
type_from_char($\n) -> lsquirly;
type_from_char($) -> rsquirly.

keyword_from_ident("fn") -> create_token(function, "fn");
keyword_from_ident("let") -> create_token(let, "let");
keyword_from_ident(_) -> undefined.

File: pascal/src/Lexer.pas
program LexerProgram;
uses SysUtils;

type
    TokenType = (Illegal, Eof, Ident, Int, Equal, Plus, Comma, Semicolon, LParen, RParen, LSquirly, RSquirly, FunctionType, Let);

type
    Token = record
        Token_Type: TokenType;
        Literal: String;
    end;

type
    Lexer = class
        Input: String;
        InputLength: Integer;
        Position: Integer;
        ReadPosition: Integer;
        Ch: Char;

        constructor Create(InputCode: String);
        procedure ReadChar();
        function GetNextToken(): Token;
        function ReadInt(): String;
        function ReadIdent(): String;
        procedure SkipWhitespace();
    end;

constructor Lexer.Create(InputCode: String);
begin
    self.Input := InputCode;
    self.InputLength := Length(InputCode);
    self.Position := 0;
    self.ReadPosition := 0;
    self.Ch := #0;
    self.ReadChar();
end;

procedure Lexer.ReadChar();
begin
    if self.ReadPosition >= self.InputLength then
        self.Ch := #0
    else
        self.Ch := self.Input[self.ReadPosition + 1];
    self.Position := self.ReadPosition;
    Inc(self.ReadPosition);
end;

function CreateToken(Token_Type: TokenType; Literal: string): Token;
var
    tok: Token;
begin
    tok.Token_Type:= Token_Type;
    tok.Literal:= Literal;

    Result:= tok;
end;

function IsAlpha(c: Char): Boolean;
begin
    Result := ((c >= 'a') and (c <= 'z')) or ((c >= 'A') and (c <= 'Z'));
end;

function IsDigit(c: Char): Boolean;
begin
    Result := (c >= '0') and (c <= '9');
end;

function IsSpace(c: Char): Boolean;
begin
    Result := (c = ' ') or (c = #9) or (c = #10) or (c = #13);
end;

function Lexer.GetNextToken(): Token;
var
    tok: Token;
    indent: String;
begin
    self.SkipWhitespace();

    if Ch = '=' then tok := CreateToken(Equal, '=') else
    if Ch = '+' then tok := CreateToken(Plus, '+') else
    if Ch = ',' then tok := CreateToken(Comma, ',') else
    if Ch = ';' then tok := CreateToken(Semicolon, ';') else
    if Ch = '(' then tok := CreateToken(LParen, '(') else
    if Ch = ')' then tok := CreateToken(RParen, ')') else
    if Ch = '{' then tok := CreateToken(LSquirly, '{') else
    if Ch = '}' then tok := CreateToken(RSquirly, '}') else
    if Ch = #0 then tok := CreateToken(Eof, #0) else
    if (IsAlpha(Ch)) or (Ch = '_') then
    begin
        indent := self.ReadIdent();
        if indent = 'fn' then tok := CreateToken(FunctionType, 'fn') else
        if indent = 'let' then tok := CreateToken(Let, 'let') else
        tok := CreateToken(Ident, indent);
    end else
    if IsDigit(Ch) then
        tok := CreateToken(Int, self.ReadInt())
    else
        tok := CreateToken(Illegal, Ch);

    self.ReadChar();
    Result := tok;
end;

function Lexer.ReadInt(): String;
var
    pos: Integer;
begin
    pos := self.Position;
    while (self.Ch >= '0') and (self.Ch <= '9') do
        self.ReadChar();
    Result := Copy(self.Input, pos + 1, self.Position - pos);
end;

function Lexer.ReadIdent(): String;
var
    pos: Integer;
begin
    pos := self.Position;
    while IsAlpha(self.Ch) or (self.Ch = '_') do
        self.ReadChar();
    Result := Copy(self.Input, pos + 1, self.Position - pos);
end;

procedure Lexer.SkipWhitespace();
begin
    while IsSpace(self.Ch) do
        self.ReadChar();
end;

function TokenTypeToString(tokenType: TokenType): String;
begin
    case tokenType of
        Illegal: Result := 'ILLEGAL';
        Eof: Result := 'EOF';
        Ident: Result := 'IDENT';
        Int: Result := 'INT';
        Equal: Result := '=';
        Plus: Result := '+';
        Comma: Result := ',';
        Semicolon: Result := ';';
        LParen: Result := '(';
        RParen: Result := ')';
        LSquirly: Result := '{';
        RSquirly: Result := '}';
        FunctionType: Result := 'FUNCTION';
        Let: Result := 'LET';
    else
        Result := '';
    end;
end;

var
    lex: Lexer;
    Input: String;
    c: Char;
begin
    Input := '=+(){},;';
    lex := Lexer.Create(Input);
    for c in Input do
    begin
        with lex.GetNextToken() do
            WriteLn('Type: ', TokenTypeToString(Token_Type), ', Literal: ', Literal);
    end;
    with lex.GetNextToken() do
        WriteLn('Type: ', TokenTypeToString(Token_Type), ', Literal: ', Literal);
end.

